File: obb.py
import numpy as np

class OBB:
    def __init__(self, center, half_sizes, orientation):
        self.center = np.array(center)
        self.half_sizes = np.array(half_sizes)
        self.orientation = np.array(orientation)  # 3x3 rotation matrix

    def intersects(self, other):
        return self.obb_intersects(self, other)

    def combine(self, other):
        points1 = self.get_corners()
        points2 = other.get_corners()
        all_points = np.vstack((points1, points2))

        new_center = np.mean(all_points, axis=0)
        new_orientation = np.eye(3)  # Simplified; you can use a more complex method to combine orientations
        new_half_sizes = (np.max(all_points, axis=0) - np.min(all_points, axis=0)) / 2

        return OBB(new_center, new_half_sizes, new_orientation)

    def get_corners(self):
        corners = []
        for dx in [-1, 1]:
            for dy in [-1, 1]:
                for dz in [-1, 1]:
                    corner = self.center + dx * self.half_sizes[0] * self.orientation[:, 0] \
                                      + dy * self.half_sizes[1] * self.orientation[:, 1] \
                                      + dz * self.half_sizes[2] * self.orientation[:, 2]
                    corners.append(corner)
        return np.array(corners)

    @staticmethod
    def obb_intersects(obb1, obb2):
        axes = np.vstack((obb1.orientation, obb2.orientation))
        axes = np.vstack((axes, np.cross(obb1.orientation, obb2.orientation)))
        axes = np.unique(axes, axis=0)  # Remove duplicate axes from cross products

        for axis in axes:
            if not OBB.overlap_on_axis(obb1, obb2, axis):
                return False
        return True

    @staticmethod
    def overlap_on_axis(obb1, obb2, axis):
        projection1 = OBB.project_obb(obb1, axis)
        projection2 = OBB.project_obb(obb2, axis)
        return projection1[0] <= projection2[1] and projection2[0] <= projection1[1]

    @staticmethod
    def project_obb(obb, axis):
        center_projection = np.dot(obb.center, axis)
        half_size_projection = np.sum(np.abs(np.dot(obb.half_sizes * obb.orientation, axis)))
        return [center_projection - half_size_projection, center_projection + half_size_projection]

File: Maman17/obb.cpp
#include "obb.h"
#include <Eigen/Dense>

bool OBB::intersects(const OBB& other) const {
    return obbIntersects(*this, other);
}

OBB OBB::combine(const OBB& other) const {
    std::vector<Eigen::Vector3f> points1 = getCorners();
    std::vector<Eigen::Vector3f> points2 = other.getCorners();
    points1.insert(points1.end(), points2.begin(), points2.end());

    Eigen::Vector3f new_center = Eigen::Vector3f::Zero();
    for (const auto& point : points1) {
        new_center += point;
    }
    new_center /= points1.size();

    Eigen::Matrix3f new_orientation = Eigen::Matrix3f::Identity(); // Simplified orientation
    Eigen::Vector3f min_point = points1[0];
    Eigen::Vector3f max_point = points1[0];

    for (const auto& point : points1) {
        min_point = min_point.cwiseMin(point);
        max_point = max_point.cwiseMax(point);
    }

    Eigen::Vector3f new_half_sizes = (max_point - min_point) / 2.0f;

    return OBB(new_center, new_half_sizes, new_orientation);
}

std::vector<Eigen::Vector3f> OBB::getCorners() const {
    std::vector<Eigen::Vector3f> corners;
    for (int dx = -1; dx <= 1; dx += 2) {
        for (int dy = -1; dy <= 1; dy += 2) {
            for (int dz = -1; dz <= 1; dz += 2) {
                Eigen::Vector3f corner = center 
                                       + dx * half_sizes[0] * orientation.col(0)
                                       + dy * half_sizes[1] * orientation.col(1)
                                       + dz * half_sizes[2] * orientation.col(2);
                corners.push_back(corner);
            }
        }
    }
    return corners;
}

bool OBB::obbIntersects(const OBB& obb1, const OBB& obb2) {
    std::vector<Eigen::Vector3f> axes;

    axes.push_back(obb1.orientation.col(0));
    axes.push_back(obb1.orientation.col(1));
    axes.push_back(obb1.orientation.col(2));
    axes.push_back(obb2.orientation.col(0));
    axes.push_back(obb2.orientation.col(1));
    axes.push_back(obb2.orientation.col(2));

    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            axes.push_back(obb1.orientation.col(i).cross(obb2.orientation.col(j)).normalized());
        }
    }

    for (const auto& axis : axes) {
        if (!overlapOnAxis(obb1, obb2, axis)) {
            return false;
        }
    }

    return true;
}

bool OBB::overlapOnAxis(const OBB& obb1, const OBB& obb2, const Eigen::Vector3f& axis) {
    auto projection1 = projectOBB(obb1, axis);
    auto projection2 = projectOBB(obb2, axis);
    return projection1.first <= projection2.second && projection2.first <= projection1.second;
}

std::pair<float, float> OBB::projectOBB(const OBB& obb, const Eigen::Vector3f& axis) {
    float center_projection = obb.center.dot(axis);
    float half_size_projection = 0.0f;

    for (int i = 0; i < 3; ++i) {
        half_size_projection += obb.half_sizes[i] * std::abs(obb.orientation.col(i).dot(axis));
    }

    return std::make_pair(center_projection - half_size_projection, center_projection + half_size_projection);
}
